python/ray/autoscaler/v2/instance_manager/reconciler.py

```diff
@@ -434,3 +490,107 @@ def _update_instance_manager(
+                    f"Ray node {n.node_id.decode()} has no instance id. "
+                    "This only happens to a ray node that's not managed by autoscaler. "
+                    "If not, please file a bug at https://github.com/ray-project/ray"
+                )
+
+        for cloud_instance_id, ray_node in ray_nodes_by_cloud_instance_id.items():
+            if cloud_instance_id not in im_instances_by_cloud_instance_id:
+                # This is a ray node that's not managed by the instance manager.
+                # or we haven't discovered the instance yet. There's nothing
+                # much we could do here.
+                logger.info(
+                    f"Ray node {ray_node.node_id.decode()} has no matching instance in "
+                    f"instance manager with cloud instance id={cloud_instance_id}."
+                )
+                continue
+
+            im_instance = im_instances_by_cloud_instance_id[cloud_instance_id]
+            reconciled_im_status = Reconciler._reconciled_im_status_from_ray_status(
+                ray_node.status, im_instance.status
+            )
+
+            if reconciled_im_status != im_instance.status:
+                updates[im_instance.instance_id] = IMInstanceUpdateEvent(
+                    instance_id=im_instance.instance_id,
+                    new_instance_status=reconciled_im_status,
+                    details="Reconciled from ray node status "
+                    f"{NodeStatus.Name(ray_node.status)} "
+                    f"for ray node {ray_node.node_id.decode()}",
+                    ray_node_id=ray_node.node_id.decode(),
+                )
+                logger.debug(
+                    "Updating {}({}) with {}.".format(
+                        im_instance.instance_id,
+                        IMInstance.InstanceStatus.Name(im_instance.status),
+                        MessageToDict(updates[im_instance.instance_id]),
+                    )
+                )
+
+        Reconciler._update_instance_manager(instance_manager, updates, version)
+
+    @staticmethod
+    def _reconciled_im_status_from_ray_status(
+        ray_status: NodeStatus, cur_im_status: IMInstance.InstanceStatus
+    ) -> "IMInstance.InstanceStatus":
+        """
+        Reconcile the instance status from the ray node status.
+        Args:
+            ray_status: the current ray node status.
+            cur_im_status: the current IM instance status.
+        Returns:
+            The reconciled IM instance status
+
+        Raises:
+            ValueError: If the ray status is unknown.
+        """
+        reconciled_im_status = None
+        if ray_status in [NodeStatus.RUNNING, NodeStatus.IDLE]:
+            reconciled_im_status = IMInstance.RAY_RUNNING
+        elif ray_status == NodeStatus.DEAD:
+            reconciled_im_status = IMInstance.RAY_STOPPED
+        elif ray_status == NodeStatus.DRAINING:
+            reconciled_im_status = IMInstance.RAY_STOPPING
```

Comment:
Is there a case where we stop without drain?

Reply:
yeah, i think for providers (Kuberay) that manage ray itself, we should just terminate the cloud instance. let me add to this PR actually.

Reply:
the preemption case should be handled in this PR: https://github.com/ray-project/ray/pull/43078